src/managed_process.rs

```diff
@@ -0,0 +1,248 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// Assumes Unix
+use libc::{self,  c_int};
+
+use std::thread;
+use std::thread::JoinHandle;
+use std::sync::{ Arc, Mutex };
+use std::process::Child;
+use std::io::{ Error, ErrorKind, Result };
+use time;
+use std::time::Duration;
+
+/// Unix exit statuses
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub struct ExitStatus(c_int);
+
+const RESTART_TIME_THRESHOLD: f64 = 5.0; // seconds
+
+pub struct ManagedProcess {
+    kill_signal: Arc<Mutex<u32>>,
+    pid:         Arc<Mutex<Option<u32>>>,
+    thread:      JoinHandle<()>
+}
+
+impl ManagedProcess {
+
+    /// Create a new ManagedProcess and start it.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tunnel_controller::ManagedProcess;
+    /// use std::process::Command;
+    ///
+    /// let process = ManagedProcess::start(|| {
+    ///     Command::new("echo")
+    ///             .arg("Hello")
+    ///             .arg("World")
+    ///             .spawn()
+    /// });
+    ///
+    /// ```
+    pub fn start<F: 'static>(spawn: F) -> Result<ManagedProcess>
+        where F: Fn() -> Result<Child> + Send {
+
+        let pid = Arc::new(Mutex::new(None));
+
+        // Uses a u32 Mutex to avoid the compiler complaining that you can use an AtomicBool.
+        // In this case we want a bool like thing _and_ a lock.
+        let kill_signal = Arc::new(Mutex::new(0));
+
+        let shared_kill_signal  = kill_signal.clone();
+        let shared_pid = pid.clone();
+
+        let thread = thread::spawn(move || {
+            // Artificial start/end time for first run
+            let mut start_time = RESTART_TIME_THRESHOLD as f64;
+            let mut end_time = 0 as f64;
+            let mut backoff = 0;
+            let mut starts = 0;
+
+            loop {
+                let mut child_process;
+
+                {
+                    let kill_signal = shared_kill_signal.lock().unwrap();
+                    let mut pid = shared_pid.lock().unwrap();
+
+                    if *kill_signal == 1 {
+                        *pid = None;
+                        debug!("Received process kill signal");
+                        break;
+                    }
+
+                    if (end_time - start_time) < RESTART_TIME_THRESHOLD {
+                       backoff += 1;
+                       let backoff_seconds = (backoff * backoff) / 2;
+                       info!("Backing off creating a new process for {} seconds", backoff_seconds);
+                       thread::sleep(Duration::new(backoff_seconds, 0));
+                    } else {
+                        backoff = 0;
+                    }
+
+                    info!("Starting process. Restarted {} times", starts);
+                    start_time = time::precise_time_s();
```

Comment:
I would rather not depend on a crate just for that. I had a similar need and did it with https://github.com/fxbox/registration_server/blob/master/src/db.rs#L65